File: ValidPalindromeII.java
/**
 * LeetCode 680 Valid Palindrome II
 */
class Solution {
    public boolean validPalindrome(String s) {
        if (s == null) {
            return false;
        }
        return palindrome(s,0,s.length() - 1,1);
    }

    public boolean palindrome(String s,int left,int right,int chance) {
        if (left > right) {
            return true;
        }
        if (s.charAt(left) == s.charAt(right)) {
            return palindrome(s,left + 1,right - 1,chance);
        } else {
            if (chance == 0) {
                return false;
            }
            return palindrome(s,left + 1,right,chance - 1) || palindrome(s,left,right - 1,chance - 1);
        }
    }
}

File: ValidPalindromeII.go
func validPalindrome(s string) bool {
	l, r := 0, len(s)-1
	for l < r {
		if s[l] == s[r] {
			l++
			r--
		} else {
			return isPalindrome(s[l:r]) || isPalindrome(s[l+1:r+1])
		}
	}
	return true
}

func isPalindrome(s string) bool {
	l, r := 0, len(s)-1
	for l < r {
		if s[l] == s[r] {
			l++
			r--
		} else {
			return false
		}
	}
	return true
}
